kolibri/core/content/utils/file_availability.py

```diff
@@ -0,0 +1,111 @@
+import json
+import os
+import re
+from itertools import compress
+
+import requests
+from django.core.cache import cache
+from django.utils.text import compress_string
+
+from kolibri.core.content.models import LocalFile
+from kolibri.core.content.utils.paths import get_content_storage_dir_path
+from kolibri.core.content.utils.paths import get_file_checksums_url
+
+
+checksum_regex = re.compile("^([a-f0-9]{32})$")
+
+
+def _filter_checksums_by_channel(checksums, channel_id):
+    """
+    Helper function to filter checksums received to what is actually used by our
+    copy of the channel. Used by both module methods, as for a disk based import
+    there could be many files present that are not part of the channel, and for
+    remote import, the remote channel might be a different version and hence
+    have different local files associated with it.
+    """
+    return list(
+        set(
+            LocalFile.objects.filter(
+                files__contentnode__channel_id=channel_id
+            ).values_list("id", flat=True)
+        ).intersection(set(checksums))
+    )
+
+
+def _generate_mask(integer_mask):
+    while integer_mask:
+        yield bool(integer_mask % 2)
+        integer_mask /= 2
+
+
+def get_available_checksums_from_remote(channel_id, baseurl):
+    CACHE_KEY = "PEER_AVAILABLE_CHECKSUMS_{baseurl}_{channel_id}".format(
+        baseurl=baseurl, channel_id=channel_id
+    )
+    if CACHE_KEY not in cache:
+
+        channel_checksums = (
+            LocalFile.objects.filter(
+                available=False, files__contentnode__channel_id=channel_id
+            )
+            .values_list("id", flat=True)
+            .distinct()
+        )
+
+        response = requests.post(
+            get_file_checksums_url(channel_id, baseurl),
+            data=compress_string(
+                bytes(json.dumps(list(channel_checksums)).encode("utf-8"))
+            ),
+            headers={"content-type": "application/gzip"},
+        )
+
+        checksums = None
+
+        # Do something if we got a successful return
+        if response.status_code == 200:
+            try:
+                integer_mask = int(response.content)
+
+                # Filter to avoid passing in bad checksums
+                checksums = list(
+                    compress(channel_checksums, _generate_mask(integer_mask))
+                )
+                cache.set(CACHE_KEY, checksums, 3600)
```

Comment:
Instead of time, could we have this get invalidated when the server's CONTENT_CACHE_KEY (or whatever it's called) changes? So, retrieve that key at the start and use it as part of this local cache key?

Reply:
See the new commend for the function to see where I landed on this. I think reducing requests is a reasonable strategy for now.